paddle/operators/load_combine_op.cc

```diff
@@ -0,0 +1,122 @@
+
+#include "paddle/framework/op_registry.h"
+#include "paddle/platform/device_context.h"
+
+namespace paddle {
+namespace operators {
+
+class LoadCombineOp : public framework::OperatorBase {
+ public:
+  LoadCombineOp(const std::string &type,
+                const framework::VariableNameMap &inputs,
+                const framework::VariableNameMap &outputs,
+                const framework::AttributeMap &attrs)
+      : OperatorBase(type, inputs, outputs, attrs) {}
+  void Run(const framework::Scope &scope,
+           const platform::Place &place) const override {
+    auto filename = Attr<std::string>("file_path");
+    auto position_counter = Attr<int>("position_counter");
+
+    std::ifstream fin(filename);
+    PADDLE_ENFORCE(static_cast<bool>(fin),
+                   "Cannot open file %s for load_combine op", filename);
+
+    auto out_var_name = Output("Out");
+    auto *out_var = scope.FindVar(out_var_name);
+    PADDLE_ENFORCE(out_var != nullptr, "Output variable %s cannot be found",
+                   out_var_name);
+
+    auto *tensor = out_var->GetMutable<framework::LoDTensor>();
+
+    platform::DeviceContextPool &pool = platform::DeviceContextPool::Instance();
+    auto &dev_ctx = *pool.Get(place);
+
+    uint64_t data_length;
+    char *buffer = NULL;
+    for (int i = 0; i <= position_counter; i++) {
+      if (!buffer) delete[] buffer;
+
+      // Error checking
+      PADDLE_ENFORCE(static_cast<bool>(fin), "Cannot read more from file %s",
+                     filename);
+
+      // Read a fixed-width int, to get the number of bytes
+      // for the serialized data.
+      fin.read(reinterpret_cast<char *>(&data_length), sizeof(data_length));
+
+      // Error checking
+      PADDLE_ENFORCE(static_cast<bool>(fin), "Cannot read more from file %s",
+                     filename);
+
+      buffer = new char[data_length];
+
+      // Read the serialized data into the buffer
+      fin.read(buffer, data_length);
+    }
+
+    std::string current_serialized_data;
+    current_serialized_data.assign(buffer, data_length);
+
+    // Create an input string stream
+    std::istringstream ist(current_serialized_data);
+    DeserializeFromStream(ist, tensor, dev_ctx);
+
+    if (!buffer) delete[] buffer;  // delete the last allocated memory
+
+    if (platform::is_gpu_place(place)) {
+      // copy CPU to GPU
+      framework::LoDTensor cpu_tensor;
+      cpu_tensor.ShareDataWith(*tensor);
+      cpu_tensor.set_lod(tensor->lod());
+
+      // reset tensor
+      out_var->Clear();
+      tensor = out_var->GetMutable<framework::LoDTensor>();
+      tensor->set_lod(cpu_tensor.lod());
+      Copy(cpu_tensor, place, dev_ctx, tensor);
+    }
+  }
+};
+
+class LoadCombineOpProtoMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  LoadCombineOpProtoMaker(OpProto *proto, OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddOutput("Out", "(Tensor) The tensor need to be load_combineed");
```

Comment:
Can we change the implementation of `load_op`? The `load_combined` operator is a little confusing.

Maybe we can change the output `Out` of `load_op` to a `Duplicable`:

```cpp
AddOutput("Out", "(Tensor) The tensor need to be loaded").AsDuplicable();
```

Then `Out` will be a vector of `Variable`s, and we can load all the `Variable`s from a file through one `load_op`.

Reply:
Or we can use `offset` in a file instead of the `position_counter`? And can we avoid to read the content of the file again and again?

Reply:
I think we can combine it with `load_op`. But in my opinion, it will look more like:
```
//book keeping code
if( combine_param ) {
   // do stuff from combine_op
} else 
  // do stuff from usual op 
}
```

Regarding your `Duplicable` suggestion: I am not sure I follow completely, but still I think we will need a deserialization logic over this merged together tensor (which maybe the same amount of computation).

Regarding your suggestion about `offset`: I think it has 2 aspects:
- If we use the offset, we will have to save it in some variable, (either make a class and make a member variable assign the offset, or something like that)
- I don't see how an offset helps much in computation saving. I think even if we have an offset, everytime we open the file to load, we still have to move the file pointer from starting to the actual location. And we still do it in this code by hopping through different chunks (and of course there is some extra overheard of `new` and copy).

Reply:
> I think we can combine it with load_op

I agree with it. Maybe can do it in next PRs.

> Regarding your suggestion about offset.  I think even if we have an offset, everytime we open the file to load, we still have to move the file pointer from starting to the actual location.

The current LoadCombineOp only process one file each time. Can it process many files at the same time? That is to say, this op loads all parameters at the same time and saves them into one file, and it doesn't need offset or point_counter. Thus, it can avoid reading the content of the file again and again.

Reply:
> That is to say, this op loads all parameters at the same time and saves them into one file, and it doesn't need offset or point_counter. Thus, it can avoid reading the content of the file again and again.

@luotao1 : Can you please explain a bit? I am sure I understand.

Reply:
To just explain my point of view a bit more, I thought that the op will be used in the manner as done currently in `save_vars()` in `fluid/io.py` ([ref](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/io.py#L63). So basically, i assumed that we will have some of usage which looks like this:

https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/io.py#L89-L97

(with `save` replaced with `save_combine`)

Reply:
> I think we can combine it with load_op

I agree.

> That is to say, this op loads all parameters at the same time and saves them into one file, and it doesn't need offset or point_counter. @luotao1 : Can you please explain a bit? 

I think what @luotao1 is saying is that you can construct a save/load program desc for loading/saving all parameters data at once. This program desc has only one load/save op that takes a list of var names as input/output and do the load/save tasks for once. This list can be obtained from the main program desc by traversing variables. By doing this, we can save a lot of time reading/write files. Thus I prefer this option.